### ISSUE TITLE
bug in dispatch

### ISSUE THREAD

**Issue by CarloLucibello** [JuliaLang/julia#18985]
The following code works correctly

``` julia
 julia> f{T<:Number}(x::T, y...) = (println(length(y)); f(y[1], y[2:end]...))
f (generic function with 1 method)

julia> f{T<:Number}(x::T) = "hi"
f (generic function with 2 methods)

julia> f(1,rand(10)...)
10
9
8
7
6
5
4
3
2
1
"hi"
```

But it fails when I change to <:Union the parametrization

``` julia
julia> f{T<:Union{Float32,Float64}}(x::T, y...) = (println(length(y)); f(y[1], y[2:end]...))
f (generic function with 1 method)

julia> f{T<:Union{Float32,Float64}}(x::T) = "hi"
f (generic function with 2 methods)

julia> f(1.,rand(10)...)
10
9
8
7
6
5
4
3
2
1
0
ERROR: BoundsError: attempt to access ()
  at index [1]
 in f(::Float64) at ./REPL[1]:1 (repeats 11 times)
 in eval_user_input(::Any, ::Base.REPL.REPLBackend) at ./REPL.jl:64
 in macro expansion at ./REPL.jl:95 [inlined]
 in (::Base.REPL.##3#4{Base.REPL.REPLBackend})() at ./event.jl:68
```

At the last step we have a wrong dispatch to the first method instead of the second

Bye,
Carlo

**Comment by nalimilan** [comment]
That's because `y...` can match zero arguments too. Change the first function to this (which is actually more natural here):
```julia
f{T<:Union{Float32,Float64}}(x::T, y, z...) = (println(length(z)+1); f(y, z...))
```

**Comment by nalimilan** [comment]
That said, a redefinition warning should probably be printed in that case.

**Comment by CarloLucibello** [comment]
isn't the vararg signature less specialized than the other one, so that f(x) should always be preferred to f(x, y...)?

Also, the fact that T<:Number works and T<:Union is clearly an incoherent behaviour (bug?)

**Comment by Gnimuc** [comment]
I'm curious: what's the subtype of `Union{Float32,Float64}`? is it necessary to use parametric method here? Why not just write `f(x::Union{Float32,Float64}, y...) = (println(length(y)); f(y[1], y[2:end]...))`?

FYI, using `typealias` will suppress the error:

```julia
julia> typealias F32F64 Union{Float32, Float64}
Union{Float32,Float64}

julia> f{T<:F32F64}(x::T) = "hi"
f (generic function with 1 method)

julia> f{T<:F32F64}(x::T, y...) = (println(length(y)); f(y[1], y[2:end]...))
f (generic function with 2 methods)

julia> f(1.,rand(10)...)
10
9
8
7
6
5
4
3
2
1
"hi"
```

**Comment by yuyichao** [comment]
I think this is a dispatch error.

This is different from the default argument case and there are only two methods and the second one should always be more specific than the first one. Of course, defining it the way @nalimilan [mentioned above](https://github.com/JuliaLang/julia/issues/18985#issuecomment-254152477) is still preferred since it's easier to handle for the compiler and is less likely to cause ambiguity when there's other methods.

**Comment by CarloLucibello** [comment]
just to point to a real case scenario, I encountered this issue here JuliaLang/Iterators.jl#85 while trying to define `iteratorsize` for chained iterators

**Comment by JeffBezanson** [comment]
I believe this is fixed by #18457 - @CarloLucibello can you confirm?

**Comment by CarloLucibello** [comment]
It's working now, nice!
```julia
julia> f{T<:Union{Float32,Float64}}(x::T, y...) = (println(length(y)); f(y[1], y[2:end]...))
f (generic function with 1 method)

julia> f{T<:Union{Float32,Float64}}(x::T) = "hi"
f (generic function with 2 methods)

julia> f(1.,rand(10)...)
10
9
8
7
6
5
4
3
2
1
"hi"
```
